examples/routing/src/router.rs

```diff
@@ -0,0 +1,172 @@
+use yew::services::routing::RouteService;
+
+use yew::prelude::worker::*;
+
+use std::collections::HashSet;
+
+use stdweb::Value;
+use stdweb::JsSerialize;
+use stdweb::unstable::TryFrom;
+
+use serde::Serialize;
+use serde::Deserialize;
+
+use std::fmt::Debug;
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize) ]
+pub struct Route<T> {
+    pub path_segments: Vec<String>,
+    pub query: Option<String>,
+    pub fragment: Option<String>,
+    pub state: T
+}
+
+impl<T> Route<T>
+    where T: JsSerialize + Clone + TryFrom<Value> + Default +'static
+{
+    pub fn to_route_string(&self) -> String {
+        let path = self.path_segments.join("/");
+        let mut path = format!("/{}", path); // add the leading '/'
+        if let Some(ref query) = self.query {
+            path = format!("{}?{}", path, query);
+        }
+        if let Some(ref fragment) = self.fragment {
+            path = format!("{}#{}", path, fragment)
+        }
+        path
+    }
+
+    pub fn current_route(route_service: &RouteService<T>) -> Self
+    {
+        let path = route_service.get_path(); // guaranteed to always start with a '/'
+        let mut path_segments: Vec<String> = path.split("/").map(String::from).collect();
+        path_segments.remove(0); // remove empty string that is split from the first '/'
+
+        let mut query: String = route_service.get_query(); // The first character will be a '?'
+        let query: Option<String> = if query.len() > 1 {
+            query.remove(0);
+            Some(query)
+        } else {
+            None
+        };
+
+        let mut fragment: String = route_service.get_fragment(); // The first character will be a '#'
+        let fragment: Option<String> = if fragment.len() > 1 {
+            fragment.remove(0);
+            Some(fragment)
+        } else {
+            None
+        };
+
+
+        Route {
+            path_segments,
+            query,
+            fragment,
+            state: T::default()
+        }
+    }
+}
+
+pub enum Msg<T>
+    where T: JsSerialize + Clone + Debug + TryFrom<Value> + 'static
+{
+    BrowserNavigationRouteChanged((String, T)),
+}
+
+
+
+impl <T> Transferable for Route<T>
+    where for <'de> T: Serialize + Deserialize<'de>
+{}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Request<T> {
+    /// Changes the route using a RouteInfo struct and alerts connected components to the route change.
+    ChangeRoute(Route<T>),
+    /// Changes the route using a RouteInfo struct, but does not alert connected components to the route change.
+    ChangeRouteNoBroadcast(Route<T>),
+    GetCurrentRoute
+}
+
+impl <T> Transferable for Request <T>
+    where for <'de> T: Serialize + Deserialize<'de>
+{}
+
+/// The Router worker holds on to the RouteService singleton and mediates access to it.
+pub struct Router<T>
+    where for <'de> T: JsSerialize + Clone + Debug + TryFrom<Value> + Default + Serialize + Deserialize<'de> + 'static
```

Comment:
These are obnoxiously long trait bounds, this is done to allow users to store and restore state objects with their history if they so choose.

I understand that they are ugly and mostly irrelevant to a user's understanding of the router, but at the same time, by allowing arbitrary specification of states, it makes the provided `Router` actor more fully featured and usable were a developer to plainly copy it to their project.